moveit_ros/hybrid_planning/trajectory_operator_plugins/src/dynamic_window_operator.cpp

```diff
@@ -0,0 +1,182 @@
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Sebastian Jahr
+ */
+
+#include <moveit/trajectory_operator_plugins/dynamic_window_operator.h>
+
+#include <moveit/kinematic_constraints/utils.h>
+
+namespace moveit_hybrid_planning
+{
+const rclcpp::Logger LOGGER = rclcpp::get_logger("local_planner_component");
+const double DWA_SCALING_FACTOR = 1.0;
+
+bool DynamicWindowOperator::initialize(const rclcpp::Node::SharedPtr& node,
+                                       const moveit::core::RobotModelConstPtr& robot_model,
+                                       const std::string& group_name)
+{
+  // Initialize cycle time based on local planner frequency
+  double local_planner_frequency = 0.0;
+  node->get_parameter<double>("local_planning_frequency", local_planner_frequency);
+  if (local_planner_frequency != 0.0)
+  {
+    cycle_time_ = 1 / local_planner_frequency;
+  }
+  else
+  {
+    RCLCPP_ERROR(LOGGER, "Unable to configure Dynamic Window operator because local planner frequency is 0");
+    return false;
+  }
+
+  reference_trajectory_ = std::make_shared<robot_trajectory::RobotTrajectory>(robot_model, group_name);
+  index_ = 0;
+  return true;
+}
+
+bool DynamicWindowOperator::addTrajectorySegment(const robot_trajectory::RobotTrajectory& new_trajectory)
+{
+  // Throw away old reference trajectory and use trajectory update
+  reference_trajectory_ = std::make_shared<robot_trajectory::RobotTrajectory>(new_trajectory);
+
+  // Parametrize trajectory and calculate velocity and accelerations
+  time_parametrization_.computeTimeStamps(*reference_trajectory_);
+
+  // Reset index
+  index_ = 0;
+  return true;
+}
+
+robot_trajectory::RobotTrajectory
+DynamicWindowOperator::getLocalTrajectory(const moveit::core::RobotState& current_state)
+{
+  // Get next desired robot state
+  moveit::core::RobotState local_trajectroy_seed = reference_trajectory_->getWayPoint(index_);
+
+  // Check if state reached
+  if (local_trajectroy_seed.distance(current_state) <= 0.01)
+  {
+    // Update index (and thus desired robot state)
+    index_ += 1;
+  }
+
+  // Calculate worst case stopping time
+  double worst_case_stop_time = calculateWorstCaseStopTime(current_state);
+
+  // Construct local trajectory
+  robot_trajectory::RobotTrajectory local_trajectory(reference_trajectory_->getRobotModel(),
+                                                     reference_trajectory_->getGroupName());
+
+  // Upper limit of the dynamic time window
+  double max_duration_from_start =
```

Comment:
I think this approach only makes sense if we are able to closely follow the desired timestamps. With non-trivial local solvers, this will most certainly not be the case. A more robust approach would be to start with the current dynamic state and iterate the waypoints forward while applying the maximum deceleration until we find a waypoint that allows coming to a stop. The scaling can be applied to either the acceleration limits or the start velocity.